upup/pkg/fi/vfs_castore.go

```diff
@@ -153,59 +156,173 @@ func (c *VFSCAStore) generateCACertificate(id string) (*keyset, *keyset, error)
 	}
 
 	t := time.Now().UnixNano()
-	serial := pki.BuildPKISerial(t)
+	serial := pki.BuildPKISerial(t).String()
 
-	keyPath := c.buildPrivateKeyPath(id, serial)
-	err = c.storePrivateKey(caPrivateKey, keyPath)
+	err = c.storePrivateKey(name, &keysetItem{id: serial, privateKey: caPrivateKey})
 	if err != nil {
 		return nil, nil, err
 	}
 
 	// Make double-sure it round-trips
-	privateKeys, err := c.loadPrivateKeys(c.buildPrivateKeyPoolPath(id))
+	privateKeys, err := c.loadPrivateKeys(c.buildPrivateKeyPoolPath(name), true)
 	if err != nil {
 		return nil, nil, err
 	}
-	if privateKeys == nil || privateKeys.primary == nil || privateKeys.primary.id != serial.String() {
+	if privateKeys == nil || privateKeys.primary == nil || privateKeys.primary.id != serial {
 		return nil, nil, fmt.Errorf("failed to round-trip CA private key")
 	}
 
-	certPath := c.buildCertificatePath(id, serial)
-	err = c.storeCertificate(caCertificate, certPath)
+	err = c.storeCertificate(name, &keysetItem{id: serial, certificate: caCertificate})
 	if err != nil {
 		return nil, nil, err
 	}
 
 	// Make double-sure it round-trips
-	certificates, err := c.loadCertificates(c.buildCertificatePoolPath(id))
+	certificates, err := c.loadCertificates(c.buildCertificatePoolPath(name), true)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	if certificates == nil || certificates.primary == nil || certificates.primary.id != serial.String() {
+	if certificates == nil || certificates.primary == nil || certificates.primary.id != serial {
 		return nil, nil, fmt.Errorf("failed to round-trip CA certifiacate")
 	}
 
 	return certificates, privateKeys, nil
 }
 
-func (c *VFSCAStore) buildCertificatePoolPath(id string) vfs.Path {
-	return c.basedir.Join("issued", id)
+func (c *VFSCAStore) buildCertificatePoolPath(name string) vfs.Path {
+	return c.basedir.Join("issued", name)
 }
 
-func (c *VFSCAStore) buildCertificatePath(id string, serial *big.Int) vfs.Path {
-	return c.basedir.Join("issued", id, serial.String()+".crt")
+func (c *VFSCAStore) buildCertificatePath(name string, id string) vfs.Path {
+	return c.basedir.Join("issued", name, id+".crt")
 }
 
-func (c *VFSCAStore) buildPrivateKeyPoolPath(id string) vfs.Path {
-	return c.basedir.Join("private", id)
+func (c *VFSCAStore) buildPrivateKeyPoolPath(name string) vfs.Path {
+	return c.basedir.Join("private", name)
 }
 
-func (c *VFSCAStore) buildPrivateKeyPath(id string, serial *big.Int) vfs.Path {
-	return c.basedir.Join("private", id, serial.String()+".key")
+func (c *VFSCAStore) buildPrivateKeyPath(name string, id string) vfs.Path {
+	return c.basedir.Join("private", name, id+".key")
 }
 
-func (c *VFSCAStore) loadCertificates(p vfs.Path) (*keyset, error) {
+func (c *VFSCAStore) parseKeysetYaml(data []byte) (*kops.Keyset, error) {
+	codecs := kopscodecs.Codecs
+	yaml, ok := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), "application/yaml")
+	if !ok {
+		glog.Fatalf("no YAML serializer registered")
```

Comment:
We need to return err or panic here, not just log.

The next line `decoder := codecs.DecoderToVersion(yaml.Serializer, kops.SchemeGroupVersion)` is going to to panic, and we already may know why we should panic or err.

Reply:
glog.Fatalf panics :-)

Reply:
DOH